Improve docs around Math/DefaultBackend & add PythonDispatcher class.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50854 Improve docs around Math/DefaultBackend & add PythonDispatcher class.**

Differential Revision: [D26008542](https://our.internmc.facebook.com/intern/diff/D26008542)

Commit:
Improve docs around Math/DefaultBackend & add PythonDispatcher class.

[ghstack-poisoned]

ailzhang commented:
Example output of python_dispatcher.py
```
# toy.register(["CPU", "Math", "AutogradCPU"])
 λ ~/dev-pytorch/tools toydispatcher python python_dispatcher.py

Computed Dispatch Table
key             kernel
---------------------------
CPU             fn_CPU [kernel]
XLA             fn_Math [math kernel]
QuantizedCPU    fn_Math [math kernel]
AutogradOther   fn_Math [math kernel]
AutogradCPU     fn_AutogradCPU [kernel]
AutogradXLA     fn_Math [math kernel]

# toy.register(["CPU", "QuantizedCPU", "Math", "AutogradCPU"])
 λ ~/dev-pytorch/tools toydispatcher python python_dispatcher.py

Computed Dispatch Table
key             kernel
---------------------------
CPU             fn_CPU [kernel]
XLA             fn_Math [math kernel]
QuantizedCPU    fn_QuantizedCPU [kernel]
AutogradOther   ambiguous_autogradother [ambiguous autogradother]
AutogradCPU     fn_AutogradCPU [kernel]
AutogradXLA     fn_Math [math kernel]

```

ezyang commented on tools/python_dispatcher.py:
```diff
@@ -0,0 +1,161 @@
+import re
+import torch._C as C
```
This is nifty. Since you're depending on torch library, I'd suggest putting this Python file inside the `torch/` folder (you can underscore it to make it private--the point is to preserve the invariant that tools Python files work without having a built PyTorch) and adding some simple tests for it so it doesn't bitrot.

ezyang reviewed (approved):
Cool! Toy is a little bit of a misnomer, because you are actually using the real live production dispatcher (similarly, there's no risk of the implementation of getting out of date.)

Commit:
Update on "Improve docs around Math/DefaultBackend & add PythonDispatcher class."




[ghstack-poisoned]

Commit:
Update on "Improve docs around Math/DefaultBackend & add PythonDispatcher class."


Differential Revision: [D26008542](https://our.internmc.facebook.com/intern/diff/D26008542)

[ghstack-poisoned]

Commit:
Update on "Improve docs around Math/DefaultBackend & add PythonDispatcher class."


Differential Revision: [D26008542](https://our.internmc.facebook.com/intern/diff/D26008542)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -286,6 +286,32 @@ In addition to the backends above, we also support the keywords:
+backward of `my_op` can no longer be derived automatically.
+
+Whether to use `Math` or `DefaultBackend` for your kernel can be decided by the following steps:
+1. Always start with a `Math` kernel that's composable of other existing operators.
```
s/composable/composed/

bhosmer replied:
Also, shouldn't we say "if you can" here? I.e. are we considering the case where they can do a backend-agnostic kernel but they're calling things that don't support autograd? 

If so, then we should so add some "if you don't want to or can't" type language to the next items as well.

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -286,6 +286,32 @@ In addition to the backends above, we also support the keywords:
     registered to `Math` key should be plain mathematical composition of other
     `at::` functions and support training and inference for any backend.
```
Here I think we should add a paragraph like:

`DefaultBackend` and `Math` keys act as defaults: for example, you can specify a custom kernel for a particular backend using a backend-specific dispatch key, and use `DefaultBackend` or `Math` to specify a generic kernel for the others.

bhosmer replied:
**Also** these aren't part of your PR, but it's worth fixing them while you're here:

line 271-272: "depending on whether or not you have CPU or CUDA tensors" should become something like "depending on which backend the passed tensors belong to" 

line 281: this link is out of date, it points to the middle of some of the new codegen. We should either update it or just give the file and a keyword to search on or something

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -286,6 +286,32 @@ In addition to the backends above, we also support the keywords:
+   you can put them in alias `Autograd` or specific keys like `AutogradCPU`.
+
 If you add `dispatch` section to any API that didn't have it before, you **have to** move
 the old implementation to `Math` field so that it's still available for other backends to use.
```
I think we need to clarify what we're telling them to do here. "move the old implementation to the `Math` field" means specifying a kernel name that wasn't there before, right? The name we generate is mostly identical to the op name, but e.g. we need to tell what to do about overload names (I think we put an underscore between the base name and the overload name, but I'm not sure) and whatever other name mangling we do.

Aside: I notice there are some ops in `native_functions.yaml` that only specify `Math` kernels right now, with the same name as the op (random examples: `broadcast_to`, `row_stack`. Are those entries necessary?

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -477,15 +503,26 @@ Here're steps to follow to decide the right dispatch keyword:
       the implementations and add dispatch section with device keywords instead.
+3. Validate the computed dispatch table matches what you want. You can use `PythonDispatcher` provided in
+[torch/_python_dispatcher.py](https://github.com/pytorch/pytorch/blob/master/torch/_python_dispacher.py).
+It shows for a certain operator, how the computed dispatch table looks like after your registrations.
```
Nit: either "how the computed dispatch table looks" or "what the computed dispatch table looks like" 😁 

This is an awesome feature though!

bhosmer commented on torch/_python_dispatcher.py:
```diff
@@ -0,0 +1,154 @@
+"""
+PythonDispatcher class is a thin python-binding to C++ dispatcher and it
+is designed to show how dispatcher precompute works. In particular,
+it shows for a certain op `foo`, how the computed dispatch table looks
```
s/how/what/

bhosmer commented on torch/_python_dispatcher.py:
```diff
@@ -0,0 +1,154 @@
+
+In the real C++ dispatcher we support many dispatch keys for different
+functionalities. For simplicity PythonDispatcher only support one example
+dispatch key in each category.
```
I'm not sure how to relate the "one example dispatch key" here to the pairs of dispatch keys below. Maybe if you said "For simplicity PythonDispatcher only supports dispatch keys for a single example of each use case. These use cases are listed below." Something like that?

bhosmer commented on torch/_python_dispatcher.py:
```diff
@@ -0,0 +1,154 @@
+    Kernels registered to this key MUST work for both inference + autograd for all backends.
+
+Note we only allow registrations to alias keys inside pytorch core library. E.g you shouldn't register
+a Math kernel from torch-xla extension, instead you should upstream the kernel into pytorch/pytorch repo.
```
Should probably say what happens if you do this by mistake 😬 

Also I don't think it's overdoing it to say "Math or DefaultBackend" here, rather than just "Math"

bhosmer reviewed (approved):
Nice work - it's especially cool to have the python thing running off the actual dispatcher!

Got a decent pile of suggested edits in here but nothing huge. Definitely ping if any of them don't make sense or seem like I've missed the point 😁

Commit:
Update on "Improve docs around Math/DefaultBackend & add PythonDispatcher class."


Differential Revision: [D26008542](https://our.internmc.facebook.com/intern/diff/D26008542)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -268,26 +268,63 @@ dispatch:
+If we already know inference kernels and derivative formulas for operators `+` and `*` in our system,
+you can just register `my_op` to `Math` and both inference & autograd will just work.
+Although it seems we only write down the inference formula here, PyTorch autograd system would correctly
+set up the backward for `my_op` using the chain formula and derivatives of `+` & `-` operators.
```
Ah sorry, missed this `-` last round, should change it to `*`

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -268,26 +268,63 @@ dispatch:
+Although it seems we only write down the inference formula here, PyTorch autograd system would correctly
+set up the backward for `my_op` using the chain formula and derivatives of `+` & `-` operators.
+In other words `d_out/d_self = 1; d_out/d_other = 2` can be derived automatically from
+the `my_op` inference kernel. Of course if we don't have derivative formula defined for either `+` or `-`,
```
same

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -268,26 +268,63 @@ dispatch:
+
+If you add backend specific kernel to any operator that only had `Math` kernel before(no dispatch section
+or only `Math` entry in dispatch section), you **have to** move the old implementation to a `<op>_math`
+kernel and put it in `Math` field so that it's still available for other backends to use.
```
I don't think we actually require a name change though, right? Sorry about my confusing  earlier comment btw - I was thinking about telling them how to reconstruct the name of the old implementation, but of course they can just look it up. How about this for the whole paragraph:

---

**Important**: because a `Math` kernel is implicitly registered for ops with no `dispatch:` section, when you add a backend-specific kernel (and hence a `dispatch:` section) to one of these, you **must** also
add a `Math:` entry that names the old kernel implementation (it's named after the op, with `_<overload name>` added if applicable), so that it's still available for other backends to use. 

---

(I was going to put something about adding a `_math` suffix for clarity, but I just checked native_functions and we don't actually do that! Most `Math:` kernels are just the op name, with a couple that put `math_` in the _front_. So maybe we shouldn't suggest anything in the docs, and just let people take their cues from what they find in native_functions. (If you think it's important for clarity we could say something here, but then we should codemod to make native_functions match, I'd say)

ailzhang replied:
SGTM! Thanks for the suggestion!

bhosmer reviewed (approved):
Awesome. Couple followup edits, main one being because of a confusing suggestion of mine in the earlier review, sorry about that 😛

Commit:
Update on "Improve docs around Math/DefaultBackend & add PythonDispatcher class."


Differential Revision: [D26008542](https://our.internmc.facebook.com/intern/diff/D26008542)

[ghstack-poisoned]

bhosmer reviewed (approved):
😍
